substrate/client/network/sync/src/engine.rs

```diff
@@ -669,8 +665,25 @@ where
 					}
 				},
 				ToServiceCommand::AnnounceBlock(hash, data) => self.announce_block(hash, data),
-				ToServiceCommand::NewBestBlockImported(hash, number) =>
-					self.new_best_block_imported(hash, number),
+				ToServiceCommand::NewBestBlockImported(hash, number) => {
+					log::debug!(target: "sync", "New best block imported {:?}/#{}", hash, number);
+
+					self.chain_sync.update_chain_info(&hash, number);
+					// TODO: remove once `SyncingEngine` is refactored
+					while let Poll::Pending = self
```

Comment:
I don't think using `FuturesUnordered` is possible because we'd need to store the entire `NotificationService` inside that future and couldn't poll anything else while that future was pending. I still don't think this is a problem but I can convert the command channel into an unbounded channel so it won't block and switch it back to bounded once the refactoring on `SyncingEngine` is complete

Reply:
An unbounded channel is also what I thought about. May be convert it now and add comment in this specific place in `SyncingEngine` to convert it back once it's possible?

Actually, I don't understand how to use the waiting version of `set_handshake` even after `SyncingEngine` is converted to `select!` without pausing the processing (assuming the channel can block).

By the way, in any case, shouldn't we evaluate `set_handshake` future once and then poll it repeatedly, instead of evaluating a new future of `set_hanshake` on every poll iteration?

Reply:
It can't be easily converted to use unbounded sender because internally the handles are converted to a stream that yields `(usize, NotificationCommand)` where `usize` denotes `SetId` which is not know by the protocol when the handle is created. I think I will add `try_set_handshake()` that doesn't block.

Reply:
Do I understand correctly that this means all the channels leading to `Notifications` would need to be converted into unbounded ones? If so, adding `try_set_handshake()` seems reasonable

Reply:
If we wish to not block on any channels towards `Notifications` which is not a bad idea, then yes they'd have to be unbounded. But on the other hand, we have to start thinking about backpressure story for `sc-network`. Having all channels unbounded is not good and having all channel bounded + `.await`ing on them is also not good so we probably have to introduce more `try_*` functions to our APIs and actually start thinking where it's OK to block and where it's not.